core/modules/qdisp/Executive.cc

```diff
@@ -372,7 +375,10 @@ void Executive::_setup() {
     if (_config.serviceUrl.compare(_config.getMockStr()) == 0) {
         _xrdSsiService = new XrdSsiServiceMock(this);
     } else {
-        _xrdSsiService = XrdSsiProviderClient->GetService(eInfo, _config.serviceUrl.c_str()); // Step 1
+        if (_xrdSsiServiceStatic == nullptr) {
+            _xrdSsiServiceStatic = XrdSsiProviderClient->GetService(eInfo, _config.serviceUrl.c_str());
+        }
+        _xrdSsiService = _xrdSsiServiceStatic; // Step 1
     }
```

Comment:
Is there a chance to run into a race condition with the static initialization in case if there will be more than one instance of `Executive`?
Perhaps getting rid of `_xrdSsiService` and replacing it with a call to the static method returning `_xrdSsiServiceStatic` would be a more reliable option:
```
# File: Executive.h
class Executive {
public:
    static XrdSsiService* getXrdSsiService() {
        std::lock_guard<std::mutex> lock(_staticInitMtx);
        if (_xrdSsiServiceStatic == nullptr) {
            _xrdSsiServiceStatic = XrdSsiProviderClient->GetService(eInfo, _config.serviceUrl.c_str());
        }
        return _xrdSsiServiceStatic;
    }

private:
    static XrdSsiService* _xrdSsiServiceStatic;
    static std::mutex _staticInitMtx;
};
Note that this will do the lazy initialization of the static pointer before its first use.

Reply:
It does need mutex, good catch. I'm not sure adding a function really improves anything, _setup is already pretty small.